src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorAdornmentManager.cs

```diff
@@ -0,0 +1,216 @@
+                    }
+
+                    list.Add(tagMappingSpan);
+                    _tagSpanToPointMap.Add(tagMappingSpan, point.Value);
+                }
+            }
+
+            return map.ToImmutable();
+        }
+
+        protected override void UpdateSpans_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection, bool removeOldTags)
+        {
+            Contract.ThrowIfNull(changedSpanCollection);
+
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            if (viewLines == null || viewLines.Count == 0)
+            {
+                return; // nothing to draw on
+            }
+
+            // removing is a separate pass from adding so that new stuff is not removed.
+            if (removeOldTags)
+            {
+                foreach (var changedSpan in changedSpanCollection)
+                {
+                    // is there any effect on the view?
+                    if (viewLines.IntersectsBufferSpan(changedSpan))
+                    {
+                        AdornmentLayer.RemoveAdornmentsByVisualSpan(changedSpan);
+                    }
+                }
+            }
+
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            foreach (var (lineNum, tagMappingSpanList) in map)
+            {
+                var tagMappingSpan = GetHighestOrderTag(tagMappingSpanList);
+                if (tagMappingSpan != null)
+                {
+                    TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span);
+                    var geometry = viewLines.GetMarkerGeometry(span);
+                    if (geometry != null)
+                    {
+                        var tag = tagMappingSpan.Tag;
+                        var classificationType = _classificationRegistryService.GetClassificationType(InlineErrorTag.TagID + tag.ErrorType);
+                        var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+                        if (!_tagSpanToPointMap.TryGetValue(tagMappingSpan, out var point))
+                        {
+                            continue;
+                        }
+
+                        var lineView = TextView.GetTextViewLineContainingBufferPosition(point);
+
+                        var visualElement = graphicsResult.VisualElement;
+                        if (tag.Location is InlineErrorsLocations.HookedToCode)
+                        {
+                            Canvas.SetLeft(visualElement, lineView.Right);
+                        }
+                        else if (tag.Location is InlineErrorsLocations.HookedToWindow)
+                        {
+                            Canvas.SetLeft(visualElement, TextView.ViewportWidth - visualElement.DesiredSize.Width);
+                        }
```

Comment:
no final else in case it's neither of tehse two values?

Reply:
it's only ever going to be one of those two values, the final else could just continue? but it should never hit it